File: clients/macos/Sources/Core/DesktopReleaseCatalog.swift
import Foundation

public struct DesktopReleaseUpdate: Equatable {
    public let tag: String
    public let assetName: String
    public let archiveURL: URL
    public let checksumURL: URL

    public init(
        tag: String,
        assetName: String,
        archiveURL: URL,
        checksumURL: URL
    ) {
        self.tag = tag
        self.assetName = assetName
        self.archiveURL = archiveURL
        self.checksumURL = checksumURL
    }
}

public enum DesktopReleaseCatalogError: LocalizedError {
    case invalidResponse

    public var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "GitHub returned an invalid desktop release catalog."
        }
    }
}

public enum DesktopReleaseCatalog {
    private static let repositoryOwner = "IPA-CyberLab"
    private static let repositoryName = "IPA-RS-HeteroNetwork"

    /// Selects the newest published release that contains both the requested
    /// archive and its checksum. Releases are ordered by GitHub's publication
    /// timestamp so prerelease tags such as `v0.1.15-dev.15` work without a
    /// second, subtly different SemVer implementation in the client.
    public static func availableUpdate(
        from data: Data,
        currentTag: String,
        assetName: String
    ) throws -> DesktopReleaseUpdate? {
        guard isSafeReleaseTag(currentTag), isKnownMacAsset(assetName) else {
            return nil
        }

        let decoded: [GitHubRelease]
        do {
            let decoder = JSONDecoder()
            decoded = try decoder.decode([GitHubRelease].self, from: data)
        } catch {
            throw DesktopReleaseCatalogError.invalidResponse
        }

        let candidates = decoded.compactMap { release -> Candidate? in
            guard !release.draft,
                  isSafeReleaseTag(release.tagName),
                  let publishedAt = parseTimestamp(release.publishedAt),
                  let archive = release.assets.first(where: { $0.name == assetName }),
                  let checksum = release.assets.first(where: { $0.name == "\(assetName).sha256" }),
                  isExpectedDownloadURL(
                      archive.browserDownloadURL,
                      tag: release.tagName,
                      assetName: assetName
                  ),
                  isExpectedDownloadURL(
                      checksum.browserDownloadURL,
                      tag: release.tagName,
                      assetName: "\(assetName).sha256"
                  )
            else {
                return nil
            }
            return Candidate(
                tag: release.tagName,
                publishedAt: publishedAt,
                archiveURL: archive.browserDownloadURL,
                checksumURL: checksum.browserDownloadURL
            )
        }

        guard let newest = candidates.max(by: {
            if $0.publishedAt == $1.publishedAt { return $0.tag < $1.tag }
            return $0.publishedAt < $1.publishedAt
        }) else {
            return nil
        }
        if newest.tag == currentTag { return nil }

        // Do not downgrade when a newer installed release is still in the
        // first page of the GitHub catalog but its assets were withdrawn.
        if let current = decoded.first(where: {
            !$0.draft && $0.tagName == currentTag
        }), let currentPublishedAt = parseTimestamp(current.publishedAt),
           currentPublishedAt >= newest.publishedAt {
            return nil
        }

        return DesktopReleaseUpdate(
            tag: newest.tag,
            assetName: assetName,
            archiveURL: newest.archiveURL,
            checksumURL: newest.checksumURL
        )
    }

    /// Selects an update from GitHub's public releases Atom feed. This is a
    /// fallback for clients whose unauthenticated API quota is exhausted.
    /// Release asset names are fixed by the release contract, and downloads
    /// remain protected by the separately published SHA-256 checksum.
    public static func availableUpdate(
        fromAtom data: Data,
        currentTag: String,
        assetName: String
    ) throws -> DesktopReleaseUpdate? {
        guard isSafeReleaseTag(currentTag), isKnownMacAsset(assetName) else {
            return nil
        }

        let entries = try AtomReleaseFeedParser.parse(data)
        let candidates = entries.compactMap { entry -> Candidate? in
            guard let publishedAt = parseTimestamp(entry.updated),
                  let tag = releaseTag(from: entry.alternateURL),
                  let archiveURL = expectedDownloadURL(tag: tag, assetName: assetName),
                  let checksumURL = expectedDownloadURL(
                      tag: tag,
                      assetName: "\(assetName).sha256"
                  )
            else {
                return nil
            }
            return Candidate(
                tag: tag,
                publishedAt: publishedAt,
                archiveURL: archiveURL,
                checksumURL: checksumURL
            )
        }

        guard let newest = candidates.max(by: {
            if $0.publishedAt == $1.publishedAt { return $0.tag < $1.tag }
            return $0.publishedAt < $1.publishedAt
        }) else {
            return nil
        }
        if newest.tag == currentTag { return nil }

        if let current = candidates.first(where: { $0.tag == currentTag }),
           current.publishedAt >= newest.publishedAt {
            return nil
        }

        return DesktopReleaseUpdate(
            tag: newest.tag,
            assetName: assetName,
            archiveURL: newest.archiveURL,
            checksumURL: newest.checksumURL
        )
    }

    private static func parseTimestamp(_ value: String?) -> Date? {
        guard let value else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let parsed = formatter.date(from: value) { return parsed }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }

    private static func isSafeReleaseTag(_ value: String) -> Bool {
        let scalars = Array(value.unicodeScalars)
        guard scalars.count > 1,
              scalars[0].value == 118,
              (48...57).contains(scalars[1].value)
        else {
            return false
        }
        return scalars.allSatisfy { scalar in
            (48...57).contains(scalar.value)
                || (65...90).contains(scalar.value)
                || (97...122).contains(scalar.value)
                || scalar.value == 46
                || scalar.value == 95
                || scalar.value == 45
        }
    }

    private static func isKnownMacAsset(_ value: String) -> Bool {
        value == "heteronetwork-client-macos-arm64.zip"
            || value == "heteronetwork-client-macos-x64.zip"
    }

    private static func isExpectedDownloadURL(
        _ url: URL,
        tag: String,
        assetName: String
    ) -> Bool {
        guard url.scheme == "https",
              url.host?.lowercased() == "github.com",
              url.user == nil,
              url.password == nil,
              url.port == nil,
              url.query == nil,
              url.fragment == nil
        else {
            return false
        }
        return url.pathComponents == [
            "/",
            repositoryOwner,
            repositoryName,
            "releases",
            "download",
            tag,
            assetName,
        ]
    }

    private static func releaseTag(from url: URL) -> String? {
        guard url.scheme == "https",
              url.host?.lowercased() == "github.com",
              url.user == nil,
              url.password == nil,
              url.port == nil,
              url.query == nil,
              url.fragment == nil,
              url.pathComponents.count == 6,
              url.pathComponents[0] == "/",
              url.pathComponents[1] == repositoryOwner,
              url.pathComponents[2] == repositoryName,
              url.pathComponents[3] == "releases",
              url.pathComponents[4] == "tag",
              isSafeReleaseTag(url.pathComponents[5])
        else {
            return nil
        }
        return url.pathComponents[5]
    }

    private static func expectedDownloadURL(tag: String, assetName: String) -> URL? {
        guard isSafeReleaseTag(tag),
              let url = URL(
                  string: "https://github.com/\(repositoryOwner)/\(repositoryName)/releases/download/\(tag)/\(assetName)"
              ),
              isExpectedDownloadURL(url, tag: tag, assetName: assetName)
        else {
            return nil
        }
        return url
    }
}

private struct GitHubRelease: Decodable {
    let tagName: String
    let publishedAt: String?
    let draft: Bool
    let assets: [GitHubReleaseAsset]

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case publishedAt = "published_at"
        case draft
        case assets
    }
}

private struct GitHubReleaseAsset: Decodable {
    let name: String
    let browserDownloadURL: URL

    enum CodingKeys: String, CodingKey {
        case name
        case browserDownloadURL = "browser_download_url"
    }
}

private struct Candidate {
    let tag: String
    let publishedAt: Date
    let archiveURL: URL
    let checksumURL: URL
}

private struct AtomReleaseEntry {
    let updated: String
    let alternateURL: URL
}

private final class AtomReleaseFeedParser: NSObject, XMLParserDelegate {
    private var entries: [AtomReleaseEntry] = []
    private var insideEntry = false
    private var updated = ""
    private var alternateURL: URL?
    private var capturesUpdated = false

    static func parse(_ data: Data) throws -> [AtomReleaseEntry] {
        let delegate = AtomReleaseFeedParser()
        let parser = XMLParser(data: data)
        parser.shouldResolveExternalEntities = false
        parser.delegate = delegate
        guard parser.parse() else {
            throw DesktopReleaseCatalogError.invalidResponse
        }
        return delegate.entries
    }

    func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        switch elementName {
        case "entry":
            insideEntry = true
            updated = ""
            alternateURL = nil
            capturesUpdated = false
        case "updated" where insideEntry:
            updated = ""
            capturesUpdated = true
        case "link" where insideEntry:
            if attributeDict["rel"] == "alternate",
               let href = attributeDict["href"] {
                alternateURL = URL(string: href)
            }
        default:
            break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if capturesUpdated {
            updated += string
        }
    }

    func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        switch elementName {
        case "updated" where insideEntry:
            capturesUpdated = false
        case "entry" where insideEntry:
            if !updated.isEmpty, let alternateURL {
                entries.append(
                    AtomReleaseEntry(
                        updated: updated.trimmingCharacters(in: .whitespacesAndNewlines),
                        alternateURL: alternateURL
                    )
                )
            }
            insideEntry = false
            capturesUpdated = false
        default:
            break
        }
    }
}

File: clients/windows/src/HeteroNetwork.Core/DesktopReleaseCatalog.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Xml;
using System.Xml.Linq;

namespace HeteroNetwork.Core;

public sealed record DesktopReleaseUpdate(
    string Tag,
    string AssetName,
    Uri ArchiveUrl,
    Uri ChecksumUrl);

public sealed class DesktopReleaseCatalogException : Exception
{
    public DesktopReleaseCatalogException(string message)
        : base(message)
    {
    }

    public DesktopReleaseCatalogException(string message, Exception innerException)
        : base(message, innerException)
    {
    }
}

public static class DesktopReleaseCatalog
{
    private const string RepositoryOwner = "IPA-CyberLab";
    private const string RepositoryName = "IPA-RS-HeteroNetwork";
    public const string WindowsAssetName = "heteronetwork-client-windows-x64.zip";

    public static DesktopReleaseUpdate? AvailableUpdate(
        ReadOnlySpan<byte> data,
        string currentTag,
        string assetName = WindowsAssetName)
    {
        if (!IsSafeReleaseTag(currentTag)
            || !string.Equals(assetName, WindowsAssetName, StringComparison.Ordinal))
        {
            return null;
        }

        GitHubRelease[] releases;
        try
        {
            releases = JsonSerializer.Deserialize<GitHubRelease[]>(data)
                ?? throw new JsonException("The release catalog was null.");
        }
        catch (JsonException error)
        {
            throw new DesktopReleaseCatalogException(
                "GitHub returned an invalid desktop release catalog.",
                error);
        }

        var candidates = new List<Candidate>();
        foreach (var release in releases)
        {
            if (release.Draft
                || !IsSafeReleaseTag(release.TagName)
                || !TryParseTimestamp(release.PublishedAt, out var publishedAt)
                || release.Assets is null)
            {
                continue;
            }

            var archive = release.Assets.FirstOrDefault(item =>
                string.Equals(item.Name, assetName, StringComparison.Ordinal));
            var checksumName = $"{assetName}.sha256";
            var checksum = release.Assets.FirstOrDefault(item =>
                string.Equals(item.Name, checksumName, StringComparison.Ordinal));
            if (archive?.BrowserDownloadUrl is null
                || checksum?.BrowserDownloadUrl is null
                || !IsExpectedDownloadUrl(
                    archive.BrowserDownloadUrl,
                    release.TagName!,
                    assetName)
                || !IsExpectedDownloadUrl(
                    checksum.BrowserDownloadUrl,
                    release.TagName!,
                    checksumName))
            {
                continue;
            }

            candidates.Add(new Candidate(
                release.TagName!,
                publishedAt,
                archive.BrowserDownloadUrl,
                checksum.BrowserDownloadUrl));
        }

        var newest = candidates
            .OrderByDescending(candidate => candidate.PublishedAt)
            .ThenByDescending(candidate => candidate.Tag, StringComparer.Ordinal)
            .FirstOrDefault();
        if (newest is null
            || string.Equals(newest.Tag, currentTag, StringComparison.Ordinal))
        {
            return null;
        }

        var current = releases.FirstOrDefault(release =>
            !release.Draft
            && string.Equals(release.TagName, currentTag, StringComparison.Ordinal));
        if (current is not null
            && TryParseTimestamp(current.PublishedAt, out var currentPublishedAt)
            && currentPublishedAt >= newest.PublishedAt)
        {
            return null;
        }

        return new DesktopReleaseUpdate(
            newest.Tag,
            assetName,
            newest.ArchiveUrl,
            newest.ChecksumUrl);
    }

    /// <summary>
    /// Selects an update from GitHub's public releases Atom feed when the
    /// unauthenticated API quota is unavailable. Asset names stay fixed by the
    /// release contract and the caller still verifies the published checksum.
    /// </summary>
    public static DesktopReleaseUpdate? AvailableUpdateFromAtom(
        ReadOnlySpan<byte> data,
        string currentTag,
        string assetName = WindowsAssetName)
    {
        if (!IsSafeReleaseTag(currentTag)
            || !string.Equals(assetName, WindowsAssetName, StringComparison.Ordinal))
        {
            return null;
        }

        XDocument document;
        try
        {
            using var stream = new MemoryStream(data.ToArray(), writable: false);
            using var reader = XmlReader.Create(stream, new XmlReaderSettings
            {
                DtdProcessing = DtdProcessing.Prohibit,
                XmlResolver = null,
                IgnoreComments = true,
                IgnoreProcessingInstructions = true,
            });
            document = XDocument.Load(reader, LoadOptions.None);
        }
        catch (XmlException error)
        {
            throw new DesktopReleaseCatalogException(
                "GitHub returned an invalid desktop release feed.",
                error);
        }

        XNamespace atom = "http://www.w3.org/2005/Atom";
        if (document.Root?.Name != atom + "feed")
        {
            throw new DesktopReleaseCatalogException(
                "GitHub returned an invalid desktop release feed.");
        }

        var candidates = new List<Candidate>();
        foreach (var entry in document.Root.Elements(atom + "entry"))
        {
            var updated = entry.Element(atom + "updated")?.Value;
            var alternate = entry.Elements(atom + "link")
                .FirstOrDefault(link =>
                    string.Equals(
                        (string?)link.Attribute("rel"),
                        "alternate",
                        StringComparison.OrdinalIgnoreCase));
            if (!TryParseTimestamp(updated, out var publishedAt)
                || !Uri.TryCreate(
                    (string?)alternate?.Attribute("href"),
                    UriKind.Absolute,
                    out var releasePage)
                || ReleaseTagFromPageUrl(releasePage) is not { } tag)
            {
                continue;
            }

            var archiveUrl = ExpectedDownloadUrl(tag, assetName);
            var checksumUrl = ExpectedDownloadUrl(tag, $"{assetName}.sha256");
            candidates.Add(new Candidate(
                tag,
                publishedAt,
                archiveUrl,
                checksumUrl));
        }

        var newest = candidates
            .OrderByDescending(candidate => candidate.PublishedAt)
            .ThenByDescending(candidate => candidate.Tag, StringComparer.Ordinal)
            .FirstOrDefault();
        if (newest is null
            || string.Equals(newest.Tag, currentTag, StringComparison.Ordinal))
        {
            return null;
        }

        var current = candidates.FirstOrDefault(candidate =>
            string.Equals(candidate.Tag, currentTag, StringComparison.Ordinal));
        if (current is not null && current.PublishedAt >= newest.PublishedAt)
        {
            return null;
        }

        return new DesktopReleaseUpdate(
            newest.Tag,
            assetName,
            newest.ArchiveUrl,
            newest.ChecksumUrl);
    }

    public static bool IsSafeReleaseTag(string? value)
    {
        if (string.IsNullOrEmpty(value)
            || value.Length < 2
            || value[0] != 'v'
            || !char.IsAsciiDigit(value[1]))
        {
            return false;
        }

        return value.All(character =>
            char.IsAsciiLetterOrDigit(character)
            || character is '.' or '_' or '-');
    }

    private static bool TryParseTimestamp(
        string? value,
        out DateTimeOffset timestamp) =>
        DateTimeOffset.TryParse(
            value,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out timestamp);

    private static bool IsExpectedDownloadUrl(Uri url, string tag, string assetName)
    {
        var expectedPath = $"/{RepositoryOwner}/{RepositoryName}/releases/download/{tag}/{assetName}";
        return url.IsAbsoluteUri
            && string.Equals(url.Scheme, Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase)
            && string.Equals(url.Host, "github.com", StringComparison.OrdinalIgnoreCase)
            && url.IsDefaultPort
            && string.IsNullOrEmpty(url.UserInfo)
            && string.IsNullOrEmpty(url.Query)
            && string.IsNullOrEmpty(url.Fragment)
            && string.Equals(url.AbsolutePath, expectedPath, StringComparison.Ordinal);
    }

    private static string? ReleaseTagFromPageUrl(Uri url)
    {
        var prefix = $"/{RepositoryOwner}/{RepositoryName}/releases/tag/";
        if (!url.IsAbsoluteUri
            || !string.Equals(url.Scheme, Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase)
            || !string.Equals(url.Host, "github.com", StringComparison.OrdinalIgnoreCase)
            || !url.IsDefaultPort
            || !string.IsNullOrEmpty(url.UserInfo)
            || !string.IsNullOrEmpty(url.Query)
            || !string.IsNullOrEmpty(url.Fragment)
            || !url.AbsolutePath.StartsWith(prefix, StringComparison.Ordinal))
        {
            return null;
        }

        var tag = url.AbsolutePath[prefix.Length..];
        return IsSafeReleaseTag(tag) ? tag : null;
    }

    private static Uri ExpectedDownloadUrl(string tag, string assetName) =>
        new($"https://github.com/{RepositoryOwner}/{RepositoryName}/releases/download/{tag}/{assetName}");

    private sealed record GitHubRelease(
        [property: JsonPropertyName("tag_name")] string? TagName,
        [property: JsonPropertyName("published_at")] string? PublishedAt,
        [property: JsonPropertyName("draft")] bool Draft,
        [property: JsonPropertyName("assets")] GitHubReleaseAsset[]? Assets);

    private sealed record GitHubReleaseAsset(
        [property: JsonPropertyName("name")] string? Name,
        [property: JsonPropertyName("browser_download_url")] Uri? BrowserDownloadUrl);

    private sealed record Candidate(
        string Tag,
        DateTimeOffset PublishedAt,
        Uri ArchiveUrl,
        Uri ChecksumUrl);
}
